### ISSUE TITLE
Render time is superlinear in the size of a paragraph

### ISSUE THREAD

**Issue by kmcallister** [Python-Markdown/markdown#161]
It's easiest to see with some slightly complex markup, a link for example.  Using the latest Python-Markdown from Git I ran:

``` .py
import time
import sys
import markdown

def time_conv(md):
    engine = markdown.Markdown()
    t0 = time.time()
    engine.convert(md)
    return time.time() - t0

link = '[link](http://google.com)'

for n in xrange(1,1000,10):
    md = ' '.join([link] * n)
    print n, time_conv(md)
    sys.stdout.flush()
```

When I plot the output in Gnuplot, it's clearly superlinear.  Taking the square root of the time produces a nice line, so I suspect it is quadratic in the size of the paragraph.

This doesn't happen when I separate the links with `'\n\n'` to make separate paragraphs.  I first noticed this problem when using single `'\n'`s with the `nl2br` extension.

I'm using Python 2.7.3 from Debian unstable.

**Comment by waylan** [comment]
I'm not sure what to make of this report.

Python-Markdown has never claimed to be fast. Just fast enough. If you want
high performance, I'd suggest a C implementation with python bindings.

That said, I suppose your results (well, what I'm assuming your results
were based on your description because you actually didn't provide the
results) make sense. Inline markup is processed one paragraph at a time. So
the regex will parse a paragraph with one link pretty quickly. However, a
single paragraph with 1,000 links will take some time. Not sure how to
speed that up aside from a complete re-factor - which is not happening any
time soon - besides I'm not even sure how to re-factor for that use case
anyway.

It appears from your report that this issue was brought on by using nl2br.
My personal opinion is that that is the worst addon to markdown ever -
which shouldn't have existed anywhere in any implementation - but that's
just me. Point is I'm not really interested in doing anything about this.
Of course, patches are always welcome.

**Comment by kmcallister** [comment]
My complaint isn't "it's slow" but rather the quadratic scaling.  In particular, this makes it harder for us to bound the rendering time by bounding the input length.

The performance with separate paragraphs is indeed "fast enough".  I would love to keep using Python-Markdown; it's remarkably flexible and we have already used that flexibility in numerous ways.  Porting all of those customizations to a C library seems like a world of pain.

You would certainly know better than I whether it is architecturally impossible to make paragraph rendering take linear time.  However I have seen code that scales quadratically due to gluing together strings in a bad order, and the fix isn't necessarily a large change.  I'll look into it further if I get a chance.  It's not obvious in principle that 1,000 links should be much slower as one paragraph than as 1,000 paragraphs.

The 1,000 links case is synthetic of course, but I discovered this through a real (and not terribly long) message in a real communications platform that takes 16 seconds to render.  In addition, we need to worry about malicious input designed to consume abnormal amounts of CPU time.  Probably we will need some CPU use limit regardless of the performance of Python-Markdown, but I don't suppose that's something that would be part of the library itself.

And I agree about `nl2br`, but it's the behavior that many users expect; users who aren't necessarily willing to sit down and read a tutorial about how Markdown works.  Of course this performance problem is not limited to `nl2br` as I demonstrated.

**Comment by waylan** [comment]
Just curious, but does this behavior also happen with other types of inline markup? For example, if you have a bunch of __ emphasized__ words rather than links. If just occurred to me that the problem may be the fact that the link regex allows for six levels of nested brackets (`[foo[bar[baz[blah[boo[etc]boo]blah]baz]bar]foo]`). A single paragraph with lots of links **might** be slowing the regex down because of that. The regex parser may be doing a lot of backtracking through the entire paragraph just to identify the first link. Each successive link in the paragraph would then do the same thing with time reduced by one. That would explain your test results. If the results are different for other types of markup, I suspect that this is the problem. It would also explain why the problem doesn't exhibit itself with multiple paragraphs (as the regex only runs on a single paragraph at a time).

Of course, if I'm right, an extension could easily swap out the regex for one that doesn't support nested brackets in links. I think we need the current behavior by default though. That said, a more efficient regex is always welcome.

**Comment by kmcallister** [comment]
Yeah, I can reproduce it with all kinds of different markup.  If I set

``` .py
link = '__here__ __are__ __some__ __emphasized__ __words__'
```

then 250 copies (i.e. 1250 words) takes more than a second to render, on a fast machine.  And again, the quadratic growth is evident from graphing the result.  (I can send you my data if you like, but I think you will have no trouble reproducing it.)

I can accept the fact that rendering lots of markup is slow.  The bug is that the slowdown is not linearly proportional to the amount of markup.  Instead it gets very slow very quickly.

**Comment by waylan** [comment]
So I ran markdown through cProfile and it looks like most of the time spent is in `treeprocessors.InlineProcessor._applyPattern()` which runs a match against a given inlinePattern (see [L230](https://github.com/waylan/Python-Markdown/blob/master/markdown/treeprocessors.py#L230)). In fact, the profile results also show a high amount of time per call to the built-in `match` method. The conclusion I came to is that some or all of the inline regex are terribly inefficient. To identify which, I threw together the following:

``` python
import markdown
import timeit

txt = '[foo](bar) '*10000
md = markdown.Markdown()

for pname, pattern in md.inlinePatterns.items():
    r = pattern.getCompiledRegExp()
    def test():
        r.match(txt)
    t = timeit.Timer('test()', 'from __main__ import test')
    print ' % 15s:' % pname, t.timeit(number=100)
```

Which results in typical output like this: 

```
        backtick: 0.851058225902
          escape: 0.372301501694
       reference: 2.24099258676
            link: 0.433806932387
      image_link: 0.369297892816
 image_reference: 0.382733238767
 short_reference: 0.449428108146
        autolink: 0.369177983172
        automail: 0.385681829634
      linebreak2: 0.394685270766
       linebreak: 0.402675955183
            html: 0.439544920504
          entity: 0.439374686729
      not_strong: 1.1404973438
       strong_em: 1.40306329303
          strong: 1.41256574908
        emphasis: 0.452382386003
       emphasis2: 1.08675148632
```

If I change the input to `txt = '__foo__ '*10000` then I get typical output like this:

```
        backtick: 0.632650389959
          escape: 0.277342126632
       reference: 0.433383350375
            link: 0.411837660645
      image_link: 0.278316421062
 image_reference: 0.294862599328
 short_reference: 0.415099945867
        autolink: 0.278749774445
        automail: 0.297756295487
      linebreak2: 0.294861965435
       linebreak: 0.350134719076
            html: 0.347042522271
          entity: 0.328506332912
      not_strong: 0.942314285476
       strong_em: 1.04910084575
          strong: 0.325295654644
        emphasis: 0.331723409588
       emphasis2: 0.893612456643
```

It is concerning that the first result is so high for the 'reference' pattern. I suspect that it is spending a lot of time backtracking on the link's brackets. Perhaps changing the order of patterns (moving 'link' ahead of 'reference') would help here.

However, what stands out in both results is that the various 'strong' and 'em' related patterns ('not_strong', strong_em', 'strong' (in the first instance), and 'emphasis2') all seem to be unusually high. The 'backtick' pattern also jumps out as a potential problem. These seem to be problems regardless of what type of input is provided.

I suppose we could try to streamline each of the above mentioned patterns. But I also wonder if part of the problem is the fact that each pattern is wrapped like this: `"^(.*?)%s(.*?)$" % pattern` (see [inlinepatterns.py L157](https://github.com/waylan/Python-Markdown/blob/master/markdown/inlinepatterns.py#L157)). I realize that uses non-greedy (`?`) matching, but I have often wondered if a better way existed (for example, using `search` rather than `match`). And for the record, I didn't write that part of the code - but it's always bothered me.

I realize the above doesn't provide answers. I'm just memorializing my research thus far. Perhaps I should redo the tests of the individual inlinePatterns with progressively larger input (like the original report) and see if any one pattern results in the same superliner run time - or if all of them do.

**Comment by waylan** [comment]
Ok, I've re-factored the test to gradually increase length of the input like in the original test and some of the patterns stand out.

``` python
import markdown
import timeit

md = markdown.Markdown()

for pname, pattern in md.inlinePatterns.items():
    r = pattern.getCompiledRegExp()
    result = [pname]
    for x in xrange(1,1000,10):
        txt = '[foo](bar) '*x
        def test():
            r.match(txt)
        t = timeit.Timer('test()', 'from __main__ import test')
        result.append(str(t.timeit(number=100)))
    print ', '.join(result)
```

With a plot of the results from the above test, most of the patterns result in a line which gradually climbs as the input length increases. However, a few patterns stand out.

There is a noticeable increase with 'backtick', and another noticeable increase with 'emphasis2' & 'not_strong' (which run in tandem), followed by another increase for 'strong' and 'strong_em' (which also run in tandem) and the a big increase for 'reference', which line runs at about 45 degrees.

If I change the input to not include any brackets (i.e., `'__foo__ '*x`), the issue with 'reference' goes away, but the rest remain (except for 'strong' (which I assume is because it matches quickly) and 'emphasis2' (I don't know why)). At least none of the lines run at 45 degrees, but there is still a noticeable (unacceptable?) increase in run time as input length is increased for the patterns: 'backtick', 'emphasis', 'not_strong', and 'strong_em' progressively.

I should note that this test does not factor in the fact that markdown continues to try to match a  given pattern until no more matches are found. My tests so far are only running the match once per input length. Of course, making that change to the test would only affect the one pattern which matches but not the rest (likely skewing the results).

In any event, it looks like the possible solutions are (in this order):
1. Explore changing order of 'reference' and 'link' patterns.
2. Improving the 'strong_em' pattern
3. Improving the 'strong' pattern
4. Improving the 'not_strong' pattern
5. Improving the 'emphasis2' pattern
6. Improving the 'emphasis' pattern
7. Improving the 'backtick' pattern

As always, patches are welcome. :-)

**Comment by kmcallister** [comment]
Thanks; that is useful information.

What's the significance of 45 degrees?  You're talking about a plot of render time versus the number of markup elements, correct?

**Comment by waylan** [comment]
This has been addressed as much as I expect we will ever do in #629. Therefore, I'm closing this.
